### ISSUE TITLE
FastModel + use_rslora=True causes gradient explosion at block_size=384 (stable at 1024, workaround: pure HF PEFT)

### ISSUE THREAD

**Issue by julien9679** [unslothai/unsloth-zoo#668]
Environment

- Unsloth 2026.4.8, Transformers 5.5.0, Torch 2.6.0+cu124
- GPU: NVIDIA RTX 3060 12GB, CUDA 8.6
- Model: google/gemma-3-4b-it
- Adapter: RS-LoRA (use_rslora=True), r=32, alpha=32

Description

When using FastModel with use_rslora=True and block_size=384, gradient norms explode catastrophically from step 1 (grad_norms 50–105) regardless of learning rate, including during warmup when LR is near zero. The same configuration at block_size=1024 is stable (grad_norms 7–11). Replacing FastModel with standard HuggingFace AutoModelForCausalLM + PEFT get_peft_model at block_size=384 produces healthy grad_norms (4–6) and converges normally to train_loss ~2.45.
The LR-independence rules out optimizer misconfiguration. The issue appears specific to FastModel — FastLanguageModel (Gemma 3 12B runs) and FastVisionModel (Gemma 4 E2B runs) did not exhibit this behaviour at comparable block sizes.

Reproduction

```
# Catastrophic — FastModel + RS-LoRA + block_size=384
model, tokenizer = FastModel.from_pretrained(
    model_name="google/gemma-3-4b-it",
    max_seq_length=384,
    load_in_4bit=False,
)
model = FastModel.get_peft_model(model, r=32, lora_alpha=32,
    use_rslora=True, use_dora=False, lora_dropout=0,
    target_modules=["q_proj","k_proj","v_proj","o_proj",
                    "gate_proj","up_proj","down_proj"])

# SFTTrainer with block_size=384, lr=5e-5, max_grad_norm=1.0
# → grad_norms 50-105 from step 1

# Stable — FastModel + RS-LoRA + block_size=1024 (same everything else)
# → grad_norms 7-11

# Stable — pure HF PEFT + RS-LoRA + block_size=384
model = AutoModelForCausalLM.from_pretrained("google/gemma-3-4b-it",
    dtype=torch.bfloat16, device_map="auto")
config = LoraConfig(r=32, lora_alpha=32, use_rslora=True,
    target_modules=[...], lora_dropout=0, bias="none",
    task_type=TaskType.CAUSAL_LM)
model = get_peft_model(model, config)
# SFTTrainer with block_size=384, lr=5e-5, max_grad_norm=1.0
# → grad_norms 4-6, train_loss converges to 2.45
```

Hypothesis

The gradient offloading optimisation ("Will smartly offload gradients to save VRAM") may interact with RS-LoRA's alpha/sqrt(r) gradient scaling in a numerically unstable way at shorter sequence lengths. At block_size=384, each optimizer step processes ~3K tokens vs ~8K at 1024. The issue does not appear with FastLanguageModel or FastVisionModel.

Workaround

Use pure HuggingFace PEFT (AutoModelForCausalLM + get_peft_model) when RS-LoRA is required with short block sizes.
